pkg/volume/csi/nodeinfomanager/nodeinfomanager.go

```diff
@@ -0,0 +1,456 @@
+				return node, false, nil
+			}
+		}
+
+		// Add/update annotation value
+		existingDriverMap[csiDriverName] = csiDriverNodeID
+		jsonObj, err := json.Marshal(existingDriverMap)
+		if err != nil {
+			return nil, false, fmt.Errorf(
+				"error while marshalling node ID map updated with driverName=%q, nodeID=%q: %v",
+				csiDriverName,
+				csiDriverNodeID,
+				err)
+		}
+
+		if node.ObjectMeta.Annotations == nil {
+			node.ObjectMeta.Annotations = make(map[string]string)
+		}
+		node.ObjectMeta.Annotations[annotationKeyNodeID] = string(jsonObj)
+
+		return node, true, nil
+	}
+}
+
+// removeNodeIDFromNode returns a function that removes node ID information matching the given
+// driver name from a Node object.
+func removeNodeIDFromNode(csiDriverName string) nodeUpdateFunc {
+	return func(node *v1.Node) (*v1.Node, bool, error) {
+		var previousAnnotationValue string
+		if node.ObjectMeta.Annotations != nil {
+			previousAnnotationValue =
+				node.ObjectMeta.Annotations[annotationKeyNodeID]
+		}
+
+		if previousAnnotationValue == "" {
+			return node, false, nil
+		}
+
+		// Parse previousAnnotationValue as JSON
+		existingDriverMap := map[string]string{}
+		if err := json.Unmarshal([]byte(previousAnnotationValue), &existingDriverMap); err != nil {
+			return nil, false, fmt.Errorf(
+				"failed to parse node's %q annotation value (%q) err=%v",
+				annotationKeyNodeID,
+				previousAnnotationValue,
+				err)
+		}
+
+		if _, ok := existingDriverMap[csiDriverName]; !ok {
+			// Value is already missing in node annotation, nothing more to do
+			return node, false, nil
+		}
+
+		// Delete annotation value
+		delete(existingDriverMap, csiDriverName)
+		if len(existingDriverMap) == 0 {
+			delete(node.ObjectMeta.Annotations, annotationKeyNodeID)
+		} else {
+			jsonObj, err := json.Marshal(existingDriverMap)
+			if err != nil {
+				return nil, false, fmt.Errorf(
+					"failed while trying to remove key %q from node %q annotation. Existing data: %v",
+					csiDriverName,
+					annotationKeyNodeID,
+					previousAnnotationValue)
+			}
+
+			node.ObjectMeta.Annotations[annotationKeyNodeID] = string(jsonObj)
+		}
+
+		return node, true, nil
+	}
+}
+
+// updateTopologyLabels returns a function that updates labels of a Node object with the given
+// topology information.
+func updateTopologyLabels(topology *csipb.Topology) nodeUpdateFunc {
+	return func(node *v1.Node) (*v1.Node, bool, error) {
+		if topology == nil || len(topology.Segments) == 0 {
+			return node, false, nil
+		}
+
+		for k, v := range topology.Segments {
+			if curVal, exists := node.Labels[k]; exists && curVal != v {
+				return nil, false, fmt.Errorf("detected topology value collision: driver reported %q:%q but existing label is %q:%q", k, v, k, curVal)
+			}
+		}
+
+		if node.Labels == nil {
+			node.Labels = make(map[string]string)
+		}
+		for k, v := range topology.Segments {
+			node.Labels[k] = v
+		}
+		return node, true, nil
+	}
+}
```

Comment:
I see this adding labels to a node, what removed those labels when the CSI driver reports different labels or is removed?

What keeps track of which labels came from csi drivers and ensures stale ones are removed from the node?

Reply:
Labels are never removed, because it could be used by multiple components, not just CSI. 

See here for more context: https://github.com/kubernetes/community/blob/7302f0a73403d493126f517cb10638f2faffc164/contributors/design-proposals/storage/container-storage-interface.md#topology-representation-in-node-objects

Here's the full design PR: https://github.com/kubernetes/community/pull/2034